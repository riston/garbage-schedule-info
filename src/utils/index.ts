import * as E from 'fp-ts/lib/Either';
import * as D from 'io-ts/Decoder';

export const unsafeGet = <I, A>(value: I, type: D.Decoder<I, A>): A => {
  const result = type.decode(value);
  if (E.isLeft(result)) {
    throw new Error(D.draw(result.left));
  }
  return result.right;
}