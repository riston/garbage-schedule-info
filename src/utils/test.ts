import pino from 'pino';
import supertest from 'supertest';
import request from 'supertest';
import * as E from 'fp-ts/lib/Either';

import { createServer } from '../server';
import { Context, TokenConfig } from '../server/type';
import { ScheduleResult } from '../fetcher/parse';
import { DateTime } from 'luxon';
interface TestContext {
  tokenConfig?: TokenConfig;
}

export const withServer = (
  context: TestContext,
  cb: (
    request: supertest.SuperTest<supertest.Test>,
    serverContext: Context,
  ) => Promise<void>,
): jest.ProvidesCallback => {
  const logger = pino();
  const tokenConfig =
    context.tokenConfig === undefined
      ? {
          xxx: {
            region: 'Virumaa',
            address: 'Test Village 4',
          },
        }
      : context.tokenConfig;

  const serverContext: Context = {
    logger,
    tokenConfig,
    fetchAndParseFn: jest.fn(() =>
      Promise.resolve(
        E.right<string, ScheduleResult>([
          {
            region: 'Rakvere linn, Lääne-Viru maakond',
            house: 'Kuuse & Kadaka 18',
            frequency_in_days: 28,
            next_visit_day: DateTime.fromISO('2021-08-16T00:00:00.000Z'),
            service_type: 'olmeprügi',
          },
        ]),
      ),
    ),
  };
  const server = createServer(serverContext, { port: 5000 });

  const app = server.getApplication();
  const service = request(app.callback());

  return async () => {
    try {
      await cb(service, serverContext);
    } catch (error) {
      fail(error);
    }
  };
};
