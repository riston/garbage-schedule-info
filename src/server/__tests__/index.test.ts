import { withServer } from '../../utils/test';

describe('Server handler testing', () => {
  it(
    'should return invalid request',
    withServer({}, async (request, context) => {
      const response = await request.get('/random');

      expect(context.fetchAndParseFn).not.toHaveBeenCalled();
      expect(response.statusCode).toBe(500);
      expect(response.body.message).toEqual('Invalid query parameters');
    }),
  );

  it(
    'request should fail with invalid token',
    withServer({}, async (request, context) => {
      const response = await request
        .get('/')
        .query({ token: 'random-invalid-token' });

      expect(context.fetchAndParseFn).not.toHaveBeenCalled();
      expect(response.statusCode).toBe(500);
      expect(response.body.message).toEqual('No matching params config found');
    }),
  );

  it(
    'should return test config with valid token',
    withServer({}, async (request, context) => {
      const response = await request.get('/').query({ token: 'xxx' });

      expect(context.fetchAndParseFn).toHaveBeenCalledWith({
        address: 'Test Village 4',
        region: 'Virumaa',
        type: undefined,
      });
      expect(response.statusCode).toBe(200);
      expect(response.body).toMatchSnapshot();
    }),
  );
});
