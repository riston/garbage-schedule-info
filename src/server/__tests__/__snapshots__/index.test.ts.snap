// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server handler testing should return test config with valid token 1`] = `
Array [
  Object {
    "frequency_in_days": 28,
    "house": "Kuuse & Kadaka 18",
    "next_visit_day": "2021-08-16T03:00:00.000+03:00",
    "region": "Rakvere linn, Lääne-Viru maakond",
    "service_type": "olmeprügi",
  },
]
`;
